### PR TITLE
API: Remove `ptp`, `itemset` and `newbyteorder` from `np.ndarray` class

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @rgommers @ngoldbaum,

This PR contains the first batch of cleanup changes for `numpy.ndarray` class described in [NEP 52 section](https://numpy.org/neps/nep-0052-python-api-cleanup.html#cleaning-up-the-niche-methods-on-numpy-ndarray). It covers removal of `ptp`, `setitem` and `newbyteorder` methods from `ndarray`. Should I add deprecation message or are these too niche to mention that?  

Please note that I left `np.matrix` and `np.ma.MaskedArray` classes intact as they are already legacy. (masked array has its own implementation and matrix class uses `np.ptp`)

### CODE REVIEW

**Comment by ngoldbaum** [comment]
> Should I add deprecation message or are these too niche to mention that?

Searching on github, there seems to be a lot of usage of `newbyteorder` to get a byteswapped version of an array. Like [here](https://github.com/pandas-dev/pandas/blob/ce5fdf0f55f47014240931a1f975f65767c2442a/pandas/io/stata.py#L1776) in pandas.

`ptp` also seems to have a decent amount of downstream usage, e.g. [here](https://github.com/scikit-image/scikit-image/blob/2ac3e141e8d2e31aa0ec10afc3a935396b0618fc/skimage/color/colorconv.py#L295) in matplotlib.

usage of `setitem` is harder to evaluate because the name isn't really unique to numpy in python code, but usages in popular downstream libraries don't immediately pop out in my search.

From that, I think we definitely at least need an `AttributeError` that tells people what the migration is when people access the removed attributes. I'm also a tiny bit nervous that the migration we need to suggest for `newbyteorder` might get complicated, judging by the documentation update. The most common usage seems to be doing `array.byteswap().newbyteorder()` to get a byteswapped copy. Is there a copy-pasteable idiom we can suggest instead that doesn't depend on details on the current dtype?

**Comment by mtsokol** [comment]
> I'm also a tiny bit nervous that the migration we need to suggest for `newbyteorder` might get complicated, judging by the documentation update. The most common usage seems to be doing `array.byteswap().newbyteorder()` to get a byteswapped copy. Is there a copy-pasteable idiom we can suggest instead that doesn't depend on details on the current dtype?

As a replacement I'm using:
```python
arr.view(arr.dtype.newbytorder(new_order))
```
as `np.ndarray.newbyteorder` points out that they are equivalent.

File: numpy/__init__.pyi
Comment (BvB93) [inline numpy/__init__.pyi:1]: FYI the `ndarray.itemset` type annotations still persist as of the moment and will have to be removed as well.

Comment (mtsokol) [reply numpy/__init__.pyi:1]: Thanks for pointing out! I removed them.

**Comment by ngoldbaum** [comment]
Let me know if you need a hand setting up the `AttributeError` in C (or just a deprecation if we decide an outright error is too disruptive).

**Comment by mtsokol** [comment]
> Let me know if you need a hand setting up the `AttributeError` in C (or just a deprecation if we decide an outright error is too disruptive).

@ngoldbaum Right, I forgot to add them. I pushed new changes - now in `ndarray` and scalar types, when accessing of each of these functions, a custom `AttributeError` is raised.

File: numpy/core/src/multiarray/getset.c [hunk 1]
```diff
@@ -938,6 +938,33 @@ array_matrix_transpose_get(PyArrayObject *self, void *NPY_UNUSED(ignored))
     return PyArray_MatrixTranspose(self);
 }
 
+static PyObject *
+array_ptp(PyArrayObject *self, void *NPY_UNUSED(ignored))
+{
+    PyErr_SetString(PyExc_AttributeError,
+                    "`ptp` has been removed from ndarray class. "
```
Comment (ngoldbaum) [inline numpy/core/src/multiarray/getset.c:945]: Minor nit, here and in the other messages I'd say "was removed from the ndarray class in NumPy 2.0".

**Review by ngoldbaum — APPROVED** [review]
It looks like matplotlib is ready. Are there any other downstream libraries we need to wait on PRs for?

Just one minor nit on the error message otherwise I think this is ready so long as downstream is ready as far as we're aware.

File: numpy/core/src/multiarray/getset.c (hunk 1 above)
Comment (mtsokol) [reply numpy/core/src/multiarray/getset.c]: Sure! I updated them.

**Comment by mtsokol** [comment]
> It looks like matplotlib is ready. Are there any other downstream libraries we need to wait on PRs for?

I checked other libraries and opened four PRs that reflect changes introduced there:
- scipy: https://github.com/scipy/scipy/pull/19253
- joblib: https://github.com/joblib/joblib/pull/1506
- pandas: https://github.com/pandas-dev/pandas/pull/55180
- scikit-learn: https://github.com/scikit-learn/scikit-learn/pull/27401

Then we're good to go.

**Comment by eendebakpt** [comment]
Both `ptp` and `itemset` I have used quite a lot. The removal of `ptp` is ok, it is a bit slower, but not too much:
```
import numpy as np
import numpy.core.multiarray as mu
x=np.random.rand(20,)
%timeit np.ptp(x)
%timeit x.ptp()
%timeit mu.maximum.reduce(x) - mu.minimum.reduce(x) # more verbose, but a fast alternative
```
results in:
```
5.2 µs ± 16.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
4.36 µs ± 181 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
2.69 µs ± 14.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
The `itemset` is a bit strange:
```
import numpy as np
x = np.random.rand(10,)
r = np.random.rand(3, 3)

%timeit x.itemset(2, 3)
%timeit x[2] = 3
%timeit r.itemset(2, 3)
%timeit r[2] = 3
%timeit r.itemset(0, 1, 3)
%timeit r[0, 1] = 3
```
results in
```
120 ns ± 3.04 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
98.9 ns ± 0.14 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
168 ns ± 18.1 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
440 ns ± 15.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
143 ns ± 4.72 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
121 ns ± 5.59 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
So the normal indexing is even faster than a direct `itemset`, even though the documentation for itemset claims "Compared to indexing syntax, `itemset` provides some speed increase ..." (in older versions of numpy/python this was true). The only exception is for indexing a 2D array with a single index. But that case might be rare.

@mtsokol The title of the PR contains `setitem`, maybe change that into `itemset`?

**Comment by mtsokol** [comment]
@eendebakpt Ah yes, I did a typo - fixed.

**Comment by ngoldbaum** [comment]
All the downstream PRs are merged so let's pull this one in. Thanks @mtsokol. As always, if you're reading this because of CI breakage and you think we made a mistake in this PR and the migrations we've offered aren't sufficient, please open an issue.

**Comment by jakirkham** [comment]
Were there deprecation warnings added to NumPy 1 for these removals somewhere?

**Comment by ngoldbaum** [comment]
> Were there deprecation warnings added to NumPy 1 for these removals somewhere?

Given that it's a major version bump we decided to do a few API cleanups without any deprecation, including these. If that proves to be too disruptive, we're open to re-adding with deprecation warnings.

**Comment by jakirkham** [comment]
We just ran into it when testing between NumPy 1 & 2

I don't think we need to readd them as there is a path to using the functions instead of the methods

Though it might be worth adding a deprecation warning in the next NumPy 1 release just so users know this is coming

**Comment by jakirkham** [comment]
cc @grlee77 (who ran into this; in case you have more thoughts here)

**Pull request merged**